Frameworks/UIKit/UIViewController.mm

```diff
@@ -1168,27 +1164,72 @@ - (void)presentViewController:(UIViewController*)controller animated:(BOOL)anima
         return;
     }
 
-    if (controller != nil) {
-        [[UIApplication sharedApplication] beginIgnoringInteractionEvents];
+    UIViewController* visibleParent = self;
+    bool shouldShow = false;
+    do {
+        // Allow presentation only if an ancestor is not NotVisible.
+        shouldShow = visibleParent->priv->_visibility != controllerNotVisible;
+    } while (!shouldShow && (visibleParent = [visibleParent parentViewController]));
+
+    if (!shouldShow) {
+        TraceWarning(TAG, L"Controller is not visible!");
+        return;
+    }
+
+    if (visibleParent->priv->_visibility != controllerVisible) {
+        TraceWarning(TAG,
+                     L"Presenting view controller %08x (%hs) on view controller %08x (%hs) which is not yet fully visible - continuing anyway.",
+                     controller,
+                     object_getClassName(controller),
+                     visibleParent,
+                     object_getClassName(visibleParent));
     }
 
+    if (visibleParent != self) {
+        TraceWarning(TAG,
+                     L"View controller %08x (%hs) not visible but ancestor %08x (%hs) is - continuing anyway.",
+                     self,
+                     object_getClassName(self),
+                     visibleParent,
+                     object_getClassName(visibleParent));
+    }
+
+    [[UIApplication sharedApplication] beginIgnoringInteractionEvents];
+
     if ([controller _hidesParent]) {
         [self _notifyViewWillDisappear:animated];
     }
 
-    priv->_modalViewController = controller;
     priv->_presentedViewController = controller;
-    if (controller != nil) {
-        [controller view];
-        controller->priv->_parentViewController = self;
-        controller->priv->_presentingViewController = self;
-        controller->priv->_presentCompletionBlock.attach([completion copy]);
+    [controller view];
+    controller->priv->_parentViewController = self;
+    controller->priv->_presentingViewController = self;
 
-        if ([controller modalPresentationStyle] == UIModalPresentationPopover) {
-            controller->priv->_popoverPresentationController.attach([[UIPopoverPresentationController alloc] initWithPresentedViewController:controller presentingViewController:self]);
-        }
+    if ([controller modalPresentationStyle] == UIModalPresentationPopover) {
+        controller->priv->_popoverPresentationController.attach([[UIPopoverPresentationController alloc] initWithPresentedViewController:controller presentingViewController:self]);
+
+        dispatch_block_t popoverPresent = ^{
+            [[UIApplication sharedApplication] endIgnoringInteractionEvents];
 
-        [controller performSelectorOnMainThread:@selector(_addToTop:) withObject:[NSNumber numberWithInt:animated] waitUntilDone:NO];
+            if (completion) {
+                completion();
+            }
```

Comment:
I _think_ even with non-ARC an implicit copy takes place as a result of the block capture (We don't have to worry about leaks because the capturing block is guaranteed to be nilled out.... so actually there are a few cases where I'm unnecessarily grabbing a strongSelf - just for safety's sake)